src/hotspot/share/opto/callnode.cpp

```diff
@@ -1515,17 +1515,24 @@ void SafePointNode::disconnect_from_root(PhaseIterGVN *igvn) {
 
 SafePointScalarObjectNode::SafePointScalarObjectNode(const TypeOopPtr* tp,
 #ifdef ASSERT
-                                                     AllocateNode* alloc,
+                                                     CallNode* alloc,
 #endif
                                                      uint first_index,
-                                                     uint n_fields) :
+                                                     uint n_fields,
+                                                     bool is_auto_box) :
   TypeNode(tp, 1), // 1 control input -- seems required.  Get from root.
   _first_index(first_index),
-  _n_fields(n_fields)
+  _n_fields(n_fields),
+  _is_auto_box(is_auto_box)
 #ifdef ASSERT
   , _alloc(alloc)
 #endif
 {
+#ifdef ASSERT
+  if (alloc != nullptr) {
+    assert(alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected alloc");
```

Comment:
Add dumping of `alloc` **before** this assert to know it during failure without rerunning test:
```
if (!alloc->is_Allocate() && !alloc->as_CallStaticJava()->is_boxing_method()) {
  alloc->dump();
}
```

Reply:
> It was my suggestion to pass call node as allocation so that we could trace back for what node SafePointScalarObject was created because you may have several Box objects for which we create SafePointScalarObject nodes.

The problem with that is `alloc` quickly becomes stale since allocations/calls are removed shortly after scalarization takes place. The only usage of `alloc()` is in `PhaseMacroExpand::scalar_replacement`:
```
assert(scobj->alloc() == alloc, "sanity");
```

Regarding the assert, frankly speaking, it looks repetitive and verbose. I'd prefer to just see it folded into:
```
assert(alloc == NULL || alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected alloc");
```

Or introduce a helper method to dump relevant info about the problematic node: 
```
assert(alloc == NULL || alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected alloc: %s", dump_node(alloc));
```

Also, `alloc == NULL` case can be eliminated by avoiding passing `NULL` in `PhaseVector::scalarize_vbox_node()`.

Reply:
Okay. Lets do as @iwanowww suggesting.

Reply:
> Okay. Lets do as @iwanowww suggesting.

Should we implement `dump_node` or should we just use `assert(alloc == NULL || alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected call node");`  without dumping?

Reply:
Creating node's dump string in separate method is complicated.
We can just use `false` in assert:
```
#ifdef ASSERT
  if (alloc != nullptr && !alloc->is_Allocate() && !alloc->as_CallStaticJava()->is_boxing_method()) {
    alloc->dump();
    assert(false, "unexpected call node for scalar replacement");
  }
#endif
```

Reply:
I did not see your latest changes before answering your question. You did exactly as I suggested. Good.

Reply:
Thank you for your review.